### ISSUE TITLE
How to save GridSearch best() and result() analyze in file ?

### ISSUE THREAD

**Issue by totoche25** [RubixML/ML#67]
Hi @andrewdalpino,

I used GridSearch to test parameters for a perceptron but I would like to save the best + the results of GridSearch.

On the documentation, this is not specified.

I specify that I saved the return of what returns the best() function in a file but there are even the matrixes which are present ...

If this is a bug or if there is something to improve behind, I would be happy to make a proposal in PR.

**Comment by andrewdalpino** [comment]
Hi @christopher-lafay thanks for the great question

You should be able to dump the returns of `best()` and `results()` to a file like something below ...

```php
$best = $estimator->best();

file_put_contents('best.json', json_encode($best, JSON_PRETTY_PRINT));
```

for JSON format, however like you mention, if one of the hyper-parameters happens to be an object, you will get the instance (including its properties such as matrices)

I think this is a good time to re-evaluate if this is the behavior we really want from these methods - or if we want them to return the class name of the object instead of the entire instance

I'm leaning in the direction of returning a class name instead

For now, you can map a callback over the params before you save them to file ...

```php
$best = array_map(function ($param) {
   return is_object($param) ? get_class($param) : $param;
}, $best);
```

I'd be happy to know your thoughts

By any chance, are you using [Grid Search](https://docs.rubixml.com/en/latest/grid-search.html) with a [Parallel](https://docs.rubixml.com/en/latest/parallel.html) backend?

P.S. Feel free to join our chat on Telegram https://t.me/RubixML

**Comment by totoche25** [comment]
Hi @andrewdalpino, thanks for your answer.

I had tried the two techniques you mentioned above, but it doesn't save the hidden layers in my file.

> {
>     "hidden_layers": {
>         "0": {},
>         "1": {},
>         "2": {},
>         "3": {},
>         "4": {},
>         "5": {}
>     },
>     "batch_size": 512,
>     "optimizer": "Rubix\\ML\\NeuralNet\\Optimizers\\Adam",
>     "alpha": 0.0001,
>     "epochs": 1000,
>     "min_change": 0.0001,
>     "window": 3,
>     "hold_out": 0.1,
>     "cost_fn": "Rubix\\ML\\NeuralNet\\CostFunctions\\CrossEntropy",
>     "metric": "Rubix\\ML\\CrossValidation\\Metrics\\FBeta"
> }

(with : `file_put_contents('best_model.json', JSON_FORCE_OBJECT|JSON_PRETTY_PRINT`)

Maybe, the idea would be to make an "AggregateReport" for GridSearch?

Otherwise, with regard to the Parralel backend, I tried to configure this but without success.

Here is the error it returns:

> In Igbinary.php line 39:
>                                                      
>   Serialization of 'class@anonymous' is not allowed

This is an error compared to save of model. (here with Igbinary serialization)

With native serialization : 

> In Native.php line 26:
>                                                      
>   Serialization of 'class@anonymous' is not allowed

I just joined the telegram, thanks! 👍

**Comment by andrewdalpino** [comment]
This is expected as `json_encode()` only serializes properties that are in scope ... All hidden layer props are protected/private. We would need to implement [jsonSerializable](https://www.php.net/manual/en/class.jsonserializable.php) to access those internal properties

As far as Backends go, the Amp backend uses anonymous classes which, unfortunately, are not compatible with the way that PHP serialization/unserialization works. This means that you will not be able to persist an estimator that has the Amp backend set. The workaround is to set the Serial backend before saving like in the example below.

![Annotation 2020-05-13 150746](https://user-images.githubusercontent.com/18690561/81860354-42339d80-952c-11ea-8a22-79c5985417ea.png)
